File: sdk-js/src/keys.js
// Key-input normalization for the public API surface.
//
// Protocol wire format is strict: lowercase 64-hex everywhere. The API
// boundary is forgiving: every place that takes a key accepts either a
// 32-byte Uint8Array/Buffer or a hex string (any case), including the
// keypair objects returned by generateEd25519()/generateX25519(), so
// callers never have to know which representation an internal layer
// wants. Normalization happens here, once, at the boundary.

import { bytesToHex, hexToBytes } from "./canonical.js";

const HEX_RE = /^[0-9a-fA-F]+$/;

/** Normalize a key to raw bytes. Accepts Uint8Array/Buffer or hex string. */
export function toRawKey(value, name, length = 32) {
  if (value instanceof Uint8Array) {
    if (value.length !== length) {
      throw new Error(`${name} must be ${length} bytes, got ${value.length}`);
    }
    return Buffer.from(value);
  }
  if (typeof value === "string") {
    if (value.length !== length * 2 || !HEX_RE.test(value)) {
      throw new Error(`${name} must be a ${length * 2}-char hex string or ${length} raw bytes`);
    }
    return Buffer.from(hexToBytes(value.toLowerCase()));
  }
  throw new Error(`${name} must be a hex string or Uint8Array, got ${typeof value}`);
}

/** Normalize a key to lowercase hex. Accepts Uint8Array/Buffer or hex string. */
export function toKeyHex(value, name, length = 32) {
  return bytesToHex(toRawKey(value, name, length));
}

/**
 * Normalize one signer. Accepts:
 *   - the object returned by generateEd25519() (role defaults to "originator")
 *   - { role?, publicKey, privateKey } with keys as hex strings or bytes
 * Returns { role, publicKey: Buffer(32), privateKey: Buffer(32) }.
 */
export function toSigner(signer, index = 0) {
  if (signer == null || typeof signer !== "object") {
    throw new Error(`signers[${index}] must be an object with publicKey and privateKey`);
  }
  const role = signer.role ?? "originator";
  if (typeof role !== "string" || role.length === 0) {
    throw new Error(`signers[${index}].role must be a non-empty string`);
  }
  const pub = signer.publicKey ?? signer.publicKeyHex;
  const priv = signer.privateKey ?? signer.privateKeyHex;
  if (pub == null || priv == null) {
    throw new Error(`signers[${index}] requires publicKey and privateKey (hex or 32 bytes)`);
  }
  return {
    role,
    publicKey: toRawKey(pub, `signers[${index}].publicKey`),
    privateKey: toRawKey(priv, `signers[${index}].privateKey`),
  };
}

/**
 * Normalize one encryption recipient. Accepts:
 *   - a hex string or 32-byte Uint8Array (the X25519 public key itself)
 *   - { publicKey } with the key as hex or bytes
 *   - the object returned by generateX25519()
 * Returns { publicKey: Buffer(32) }.
 */
export function toRecipient(recipient, index = 0) {
  const value =
    recipient != null && typeof recipient === "object" && !(recipient instanceof Uint8Array)
      ? (recipient.publicKey ?? recipient.publicKeyHex)
      : recipient;
  if (value == null) {
    throw new Error(`recipients[${index}] requires a publicKey (hex or 32 bytes)`);
  }
  return { publicKey: toRawKey(value, `recipients[${index}].publicKey`) };
}

/** Current UTC time as an ISO 8601 string with second precision. */
export function nowIso() {
  return new Date().toISOString().replace(/\.\d+Z$/, "Z");
}

File: sdk-py/src/capsule/keys.py
"""Key-input normalization for the public API surface. Mirrors sdk-js/src/keys.js.

Protocol wire format is strict: lowercase 64-hex everywhere. The API
boundary is forgiving: every place that takes a key accepts either 32
raw bytes or a hex string (any case), including the keypair objects
returned by ``generate_ed25519()`` / ``generate_x25519()``, so callers
never have to know which representation an internal layer wants.
Normalization happens here, once, at the boundary.
"""

from __future__ import annotations

import re
from datetime import UTC, datetime
from typing import Any

from .canonical import bytes_to_hex

KeyLike = str | bytes | bytearray | memoryview

_HEX_RE = re.compile(r"^[0-9a-fA-F]+$")


def to_raw_key(value: Any, name: str, length: int = 32) -> bytes:
    """Normalize a key to raw bytes. Accepts bytes-like or a hex string."""
    if isinstance(value, (bytes, bytearray, memoryview)):
        raw = bytes(value)
        if len(raw) != length:
            raise ValueError(f"{name} must be {length} bytes, got {len(raw)}")
        return raw
    if isinstance(value, str):
        if len(value) != length * 2 or not _HEX_RE.match(value):
            raise ValueError(f"{name} must be a {length * 2}-char hex string or {length} raw bytes")
        return bytes.fromhex(value.lower())
    raise ValueError(f"{name} must be a hex string or bytes, got {type(value).__name__}")


def to_key_hex(value: Any, name: str, length: int = 32) -> str:
    """Normalize a key to lowercase hex. Accepts bytes-like or a hex string."""
    return bytes_to_hex(to_raw_key(value, name, length))


def _field(obj: Any, *names: str) -> Any:
    """Read the first present field from a dict or attribute-style object."""
    for n in names:
        if isinstance(obj, dict):
            if obj.get(n) is not None:
                return obj[n]
        else:
            value = getattr(obj, n, None)
            if value is not None:
                return value
    return None


def to_signer(signer: Any, index: int = 0) -> dict:
    """Normalize one signer.

    Accepts the ``Ed25519KeyPair`` returned by ``generate_ed25519()``
    (role defaults to "originator") or a dict of
    ``{"role"?, "public_key", "private_key"}`` with keys as hex strings
    or bytes. Returns ``{"role", "public_key": bytes, "private_key": bytes}``.
    """
    if signer is None or isinstance(signer, (str, bytes, bytearray, memoryview)):
        raise ValueError(
            f"signers[{index}] must be a keypair or dict with public_key and private_key"
        )
    role = _field(signer, "role") or "originator"
    if not isinstance(role, str) or not role:
        raise ValueError(f"signers[{index}].role must be a non-empty string")
    pub = _field(signer, "public_key", "public_key_hex")
    priv = _field(signer, "private_key", "private_key_hex")
    if pub is None or priv is None:
        raise ValueError(f"signers[{index}] requires public_key and private_key (hex or 32 bytes)")
    return {
        "role": role,
        "public_key": to_raw_key(pub, f"signers[{index}].public_key"),
        "private_key": to_raw_key(priv, f"signers[{index}].private_key"),
    }


def to_recipient(recipient: Any, index: int = 0) -> bytes:
    """Normalize one encryption recipient to its raw 32-byte X25519 public key.

    Accepts a hex string, raw bytes, a dict with ``public_key``, or the
    ``X25519KeyPair`` returned by ``generate_x25519()``.
    """
    value = (
        recipient
        if isinstance(recipient, (str, bytes, bytearray, memoryview))
        else _field(recipient, "public_key", "public_key_hex")
    )
    if value is None:
        raise ValueError(f"recipients[{index}] requires a public_key (hex or 32 bytes)")
    return to_raw_key(value, f"recipients[{index}].public_key")


def now_iso() -> str:
    """Current UTC time as an ISO 8601 string with second precision."""
    return datetime.now(UTC).strftime("%Y-%m-%dT%H:%M:%SZ")
